pkg/networkpath/traceroute/runner.go

```diff
@@ -135,45 +171,106 @@ func (r *Runner) RunTraceroute(ctx context.Context, cfg Config) (payload.Network
+
+	return pathResult, nil
+}
+
+func (r *Runner) runTCP(cfg Config, hname string, target net.IP, maxTTL uint8, timeout time.Duration) (payload.NetworkPath, error) {
+	destPort := cfg.DestPort
+	if destPort == 0 {
+		destPort = 80 // TODO: is this the default we want?
+	}
 
-	pathResult, err := r.processResults(results, hname, rawDest, destPort, dest)
+	tr := tcp.TCPv4{
+		Target:   target,
+		DestPort: destPort,
+		NumPaths: 1,
+		MinTTL:   uint8(DefaultMinTTL),
+		MaxTTL:   maxTTL,
+		Delay:    time.Duration(DefaultDelay) * time.Millisecond,
+		Timeout:  timeout,
+	}
+
+	results, err := tr.TracerouteSequential()
 	if err != nil {
 		return payload.NetworkPath{}, err
 	}
-	log.Debugf("Processed Results: %+v", results)
+	log.Debugf("TCP Results: %+v", results)
+
+	pathResult, err := r.processTCPResults(results, hname, cfg.DestHostname, destPort, target)
+	if err != nil {
+		return payload.NetworkPath{}, err
+	}
+	log.Debugf("Parsed TCP Results: %+v", pathResult)
 
-	// TODO: better tagging
-	tracerouteRunnerTelemetry.runs.Inc()
 	return pathResult, nil
 }
 
-func getPorts(configDestPort uint16) (uint16, uint16, bool) {
-	var destPort uint16
-	var srcPort uint16
-	var useSourcePort bool
-	if configDestPort > 0 {
-		// Fixed Destination Port
-		destPort = configDestPort
-		useSourcePort = true
-	} else {
-		// Random Destination Port
-		destPort = DefaultDestPort + uint16(rand.Intn(30))
-		useSourcePort = false
+func (r *Runner) processTCPResults(res *tcp.Results, hname string, destinationHost string, destinationPort uint16, destinationIP net.IP) (payload.NetworkPath, error) {
+	pathID := uuid.New().String()
+	traceroutePath := payload.NetworkPath{
+		PathID:    pathID,
+		Timestamp: time.Now().UnixMilli(),
+		Source: payload.NetworkPathSource{
+			Hostname:  hname,
+			NetworkID: r.networkID,
+		},
+		Destination: payload.NetworkPathDestination{
+			Hostname:  getDestinationHostname(destinationHost),
+			Port:      destinationPort,
+			IPAddress: destinationIP.String(),
+		},
 	}
-	srcPort = DefaultSourcePort + uint16(rand.Intn(10000))
-	return destPort, srcPort, useSourcePort
+
+	// get hardware interface info
+	//
+	// TODO: using a gateway lookup may be a more performant
+	// solution for getting the local addr to use
+	// when sending traceroute packets in the TCP implementation
+	// really we just need the router piece
+	// might be worth also looking in to sharing a router between
+	// the gateway lookup and here or exposing a local IP lookup
+	// function
+	if r.gatewayLookup != nil {
+		src := util.AddressFromNetIP(res.Source)
+		dst := util.AddressFromNetIP(res.Target)
+
+		traceroutePath.Source.Via = r.gatewayLookup.LookupWithIPs(src, dst, r.nsIno)
+	}
+
+	for i, hop := range res.Hops {
+		isSuccess := false
+		hopname := fmt.Sprintf("unknown_hop_%d", i)
+		hostname := hopname
+
+		if !hop.IP.Equal(net.IP{}) {
+			isSuccess = true
+			hopname = hop.IP.String()
+			hostname = getHostname(hop.IP.String())
```

Comment:
Again, I feel that this is a fine to start (especially since we do getHostname in the UDP version), but do we want the system-probe to be doing hostname resolution? This feels like something that can be cached, and adding another cache to the system-probe worries me.

Reply:
We may not have a choice when it comes to system-probe doing this reverse lookup. There is no caching in the `net` library for this, so a cache would have to be implemented here. In the absence of a local cache, there is a good chance the local DNS resolver is caching, so this may not be a time consuming operation.

Reply:
I would like to see a timeout on this operation though, and I think this is already listed as a TODO for the `getHostname` implementation.

Reply:
Any opinion on what a good timeout for this would be? We call the function pretty frequently. 100ms? 500ms? 1s?

Reply:
Added a timeout for this